### ISSUE TITLE
Mix-review: Alleviate Forgetting in the Pretrain-Finetune Framework for Neural Language Generation Models

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#7]
# 논문 정보 (URL, 저자 등)

[https://arxiv.org/abs/1910.07117](https://arxiv.org/abs/1910.07117)

MIT, Facebook AI Research

# 개요

- 본 논문은 대규모의 사전학습- 파인튜닝 프레임 워크가 어떻게 동작하는지를 분석한 논문입니다.
    - 분석 결과, transformer encoder-decoder 모델은 대규모 사전학습 동안 배운 중요한 언어 생성 스킬을 파인 튜닝 시에 까먹는 경향이 있다는 사실을 발견했습니다.
    - 본 논문에서는 **까먹는 현상**을 **context sensitivity**와 **knowledge transfer**의 관점에서 분석해보았습니다.
- 이 현상을 완화시키기 위해 사전 학습용 데이터와 파인튜닝용 데이터를 혼합하는 **Mix-Review** 전략을 제안합니다.
- 본 논문의 contribution은 다음과 같습니다.
    1. **Data Separation** (사전학습, 파인튜닝 데이터 사이의 중복이 거의 없는 상태)는 모델이 중요한 언어 생성 스킬을 까먹는 원인이라는 것을 발견했습니다.
    2. **Data Mixing** 전략이 파인 튜닝 프로세스를 효과적으로 regularize하고, 까먹는 현상이 크게 완화되는 것을 확인했습니다.
    3. 실제 예시를 통해 대화 모델의 결과와 Data Mixing 전략의 영향을 확인했습니다.

# 모델

- End-to-End 답변 생성 모델은 seq2seq 모델로 공식화할 수 있습니다.
    - 여기서는 transformer encoder-decoder 구조를 이용합니다.
    - 설정은 Attention is all you need와 동일한 6 layers, 16 attention head, embedding dimension 1024, feed-forward dimension 4096을 이용합니다.
- 학습은 아래 식과 같은 negative log-likelihood를 최소화하는 식으로 진행됩니다.
<img width="611" alt="스크린샷 2020-08-31 오후 3 59 43" src="https://user-images.githubusercontent.com/49383323/91691670-ff610880-eba2-11ea-840b-8f7f29544a19.png">

- 디코딩 전략으로는 top-K sampling 방식을 이용하고, k=30으로 설정했습니다.

## 사전 학습

- 본 논문에서는 2가지 사전 학습 전략을 비교합니다.
1. Next-Sentence pre-training (NS)
    - Encoder에 context 입력하고, decoder에서 reply를 출력하는 구조
2. masked sequence-to-sequence pre-training (MASS)
    - 랜덤하게 context의 segment를 마스킹하고, decoder 쪽에서 masked segment를 생성하는 구조
- 사전 학습 과정에서 얻을 수 있는 생성 능력은 크게 2가지입니다.
    1. 지식의 획득
        - 대규모 사전학습 데이터는 대량의 지식을 담고 있고, 이러한 지식은 답변이 더 informative하고 engaging하게 만들어줍니다.
    2. 문맥 정보의 활용
        - 하지만, 현재의 많은 오픈 도메인 대화 모델은 문맥 정보에 둔감하기 때문에 뻔한 답변 문제가 발생합니다.

## 파인 튜닝

- CCNEWS 데이터로 사전학습하고, 대화 데이터로 파인튜닝을 했을 때의 Negative log-likelihood 그래프
    1. Overfitting: 학습용 대화 데이터의 NLL 로스는 계속 떨어지는 반면, 검증용 대화 데이터의 NLL 로스는 점점 증가합니다.
    2. Forgetting: 학습용으로 사용했던 CCNEWS 검증용 데이터의 로스가 점점 증가합니다.
        - 사실 타겟 데이터에 대한 성능이 좋으면 그만이긴 하지만, 모델이 사전 학습 때 배운 생성 스킬을 까먹었다고 의심할 수 있습니다.

<img width="326" alt="스크린샷 2020-08-31 오후 4 00 51" src="https://user-images.githubusercontent.com/49383323/91691802-3b946900-eba3-11ea-9c58-2305b94857f4.png">

**Mix-Review**

- 본 논문에서 제안하는 Mix-Review는 타겟 대화 데이터와 사전 학습 데이터를 일부 섞는 방법입니다.
    - mix-ratio: 사전 학습용 데이터를 얼마나 섞을지에 대한 하이퍼파라미터
    - mix-decay: 에폭이 진행됨에 따라 사전학습 데이터의 양을 얼마나 줄일지에 대한 하이퍼파라미터
- 예를 들어 100만개의 대화 데이터로 파인튜닝을 한다고 하고 mix-ratio=4, mix-decay=0.9라고 하면
    - 400만개의 사전학습 데이터를 섞어서 파인튜닝하고, 2에폭 때는 360만건으로 줄이는 느낌

<img width="516" alt="스크린샷 2020-08-31 오후 4 02 04" src="https://user-images.githubusercontent.com/49383323/91691849-5535b080-eba3-11ea-9908-62a319158c1b.png">

- mix-review의 효과를 확인하기 위해 L2 regularization을 했을때와 로스를 비교해보았을 때
    - 어느정도 regularization 효과가 있긴하지만, mix-review만큼 강력하지는 않았습니다.

<img width="310" alt="스크린샷 2020-08-31 오후 4 02 26" src="https://user-images.githubusercontent.com/49383323/91691878-61217280-eba3-11ea-9fc3-de1aad390800.png">

- 추가적으로 regularization 테크닉들로
    - dropout rate를 올리기, bottom layers를 freeze하기 등을 시도해보았지만, 큰 효과는 없었습니다.
    - 저희는 그 이유가 트랜스포머 모델이 이미 well-tuned 모델이기 때문이라고 믿고 있습니다. (드랍 아웃이나, layer norm 등의 효과가 이미 잘 녹아있는 느낌)

# 데이터

**사전학습용 데이터**

- CommonCrawl 뉴스 데이터셋 중 영어 데이터인 CCNEWS data를 사용했습니다.
    - 전체 셋은 10억 문장 (270억 단어)이나, 각종 실험을 시간내에 수행하기 위해 전체 중 10%만 이용했습니다.

**파인튜닝용 데이터**

- Dailydialogue (130만 단어) , Switchboard (120만 단어), Cornell Movie (450만 단어) 3가지 오픈 도메인 데이터셋을 이용했습니다.

**Vocabulary**

- CCNEWS-100만 데이터로 BPE를 학습해, 62k의 vocab을 생성했고, 파인튜닝시에도 이 vocab을 그대로 이용했습니다.

# 실험

- **실험 설정**
    - Adam Optimizer
    - batch size:2048
    - learning rate 0.0001 with inverse square root LR scheduler
    - half-precision(float16) pre-training
    - fine-tuning 시에 early stopping
- 실험에서는 모델이 fine-tuning 중 언어 생성 스킬을 잊어버리지 않았는지, mix-review가 언어 생성 스킬을 기억하는데 어떠한 도움을 주는지를 확인하는 것을 목표로 했습니다.
- 평가를 할땐 AMT를 이용하여 생성된 답변이 fluency, consistency, engagingness 관점에서 얼마나 좋은 답변인지 1-9점 사이로 레이블링하는 방식으로 평가를 진행했습니다.
    - 총 2500개의 데이터 페어를 레이블링받았고, human evaluation 이외의 BLEU 등의 automatic metric은 top-k sampling 방식에 적합하지 않아 사용하지 않았습니다.

<img width="663" alt="스크린샷 2020-08-31 오후 4 03 00" src="https://user-images.githubusercontent.com/49383323/91691922-77c7c980-eba3-11ea-8fe2-7460cc6b557e.png">

- 테이블 2 분석 결과
    - 우선, from scratch보다는 pretrain을 했을 때 압도적으로 좋은 성능을 보여주고 있습니다.
    - 또한, MASS보다는 NS 사전학습이 더 좋은 성능을 보여주고 있습니다.
    - 다만, mix-review와 WD사이에는 큰 차이가 없는데, 이것은 mix-review의 strong regularization effect 때문이라고 추정하고 있습니다.
        - 따라서, 좀 더 정교한 regularization 트릭이 필요하지 않을까 생각합니다.

<img width="636" alt="스크린샷 2020-08-31 오후 4 03 30" src="https://user-images.githubusercontent.com/49383323/91691973-9037e400-eba3-11ea-8458-a866feae2b47.png">

- 테이블 3에서 주목할만한 건 mix-review 방식의 생성 결과, "belgian ale" 같은 맥주와 관련된 복잡한 단어가 생성되었다는 점인데, 이 단어는 파인튜닝용 데이터에는 존재하지 않았던 단어입니다.
    - 이는, mix-review로 파인튜닝했을 때, 사전학습에서 얻은 지식을 유효하게 활용하고 있다는 것을 의미합니다.

### Behavior Analysis: Context Sensitivity

- 모델의 Context Sensitivity는 생성 모델의 중요한 속성중 하나이기 때문에, 각 모델에 따른 Context Sensitivity를 분석했습니다.
    - 문맥을 왜곡시키기 위해 (1) 문맥에서 단어의 30%를 랜덤하게 삭제 (2) 문맥에서 단어를 무작위로 shuffle해보았습니다.

<img width="643" alt="스크린샷 2020-08-31 오후 4 04 05" src="https://user-images.githubusercontent.com/49383323/91692039-af367600-eba3-11ea-8784-86ca65f83fc6.png">

- 우선, scratch부터 학습한 모델은 문맥에 둔감하다는 사실을 확인했고, 사전학습을 진행한 모델은 문맥의 변화에 따라 PPL의 변화 폭이 큰 것을 발견했습니다.
- 또한, NS 사전학습 대화 모델 (3번째 행)은 파인튜닝을 하지 않은 사전 학습 모델 (1번째 행)보다 문맥 변화에 둔감하다는 사실을 발견했습니다.
    - 이 사실은 파인 튜닝 도중 중요한 생성 스킬을 망각했다는 것을 의미합니다.
- 반면, mix-review 방식은 문맥 변화에 민감하다는 사실을 확인했습니다.

### Behavior Analysis: Knowledge Transfer

- 여기서는 모델이 얼마나 많은 지식을 갖고 있는지 확인할 수 있는 프로세스를 설계하고, 사전학습-파인튜닝 프레임워크에서 이러한 지식이 어떻게 전이되는지를 확인합니다.
- 사전학습용 데이터인 CCNEWS는 공공 뉴스 데이터이기 때문에, 이 모델이 **"big news"**에 대한 지식을 갖고 있을 거라 기대하고, 2016년 Google trend data를 활용했습니다.
    - 여기에는 365개의 trending terms (e.g. iPhone7, Deadpool 등)이 포함되어 있습니다.
- 모델이 특정 단어에 대한 지식을 갖고 있는지 확인하기 위해, 모델에 "trigger templates"를 각각 뉴스 스타일과 대화 스타일로 입력했습니다.
    - 그 후, 생성된 답변과 Ground Truth 간의 BLEU 스코어를 계산했습니다.
<img width="647" alt="스크린샷 2020-08-31 오후 4 05 03" src="https://user-images.githubusercontent.com/49383323/91692088-c5443680-eba3-11ea-8106-e8fd41dd6aa2.png">

<img width="641" alt="스크린샷 2020-08-31 오후 4 04 12" src="https://user-images.githubusercontent.com/49383323/91692066-bfe6ec00-eba3-11ea-9251-70ba2361047f.png">

- 일단, NS Pre-trained 모델의 결과를 보면 뉴스 트리거 문장의 성능이 더 높습니다.
- NS+finetune 모델은 Baseline보다는 높지만, NS Pre-trained 보다는 낮습니다.
    - 이는 망각 효과를 의미합니다.
- NS+ Mix-review는 지식을 효과적으로 보존하고 있는 것을 확인했습니다.
    - **또한 Dialogue Trigger의 문장을 입력해도 답변이 지식을 담고 있다는 것을 알 수 있습니다.**

<img width="634" alt="스크린샷 2020-08-31 오후 4 05 25" src="https://user-images.githubusercontent.com/49383323/91692122-cd03db00-eba3-11ea-8571-31894160f78a.png">

- AMT로 정성 평가를 해봤을 때도 NS+ Mix-reviews는 전체적으로 좋은 성능을 보여주고 있습니다.

### Implications and Discussion

<img width="503" alt="스크린샷 2020-08-31 오후 4 05 46" src="https://user-images.githubusercontent.com/49383323/91692153-ddb45100-eba3-11ea-91ca-7f56a1169af6.png">

- UMAP projection이라는 방법을 통해 각 모델을 클러스터링했을 때, mix-review 모델은 pretrained model과 거리가 가깝다는 것을 확인할 수 있었습니다.
- 오픈 도메인 대화에서, context sensitivity와 knowledge transfer 능력은 data-driven knowledge가 풍부한 대화를 할 수 있는 가능성을 암시합니다.

<img width="638" alt="스크린샷 2020-08-31 오후 4 06 36" src="https://user-images.githubusercontent.com/49383323/91692211-f4f33e80-eba3-11ea-9f50-6f832fafc5e5.png">

- 싱글턴, 멀티턴 대화에서 흥미로운 답변을 한 예시입니다. 특히, 개인적인 의견을 만들어 유저에게 충고를 해주는 예시가 인상적.
- 다만, 악의적인 답변을 하기도 합니다.
    - 학습에 이용한 Dailydialogue 대화셋은 모두 polite한 세션으로만 구성이 되어있어, scratch부터 학습한 모델은 이런 악의적인 답변을 할 수 없지만, 사전학습에서 배운 지식이 있어 악의적인 답변도 할 수 있게 되었습니다.

# 논의할만한 내용

# 관련된 논문
